app/src/main/java/com/kpiroom/bubble/ui/login/LoginLogic.kt

```diff
@@ -1,22 +1,138 @@
     val password = MutableLiveData<String>()
+    val confirmPassword = MutableLiveData<String>()
 
-    private fun signUpOrSignIn() {
-        //TODO("Implement db request for signin or signup")
-        //login.value
-        //password.value
+    val authButtonText = MutableLiveData<String>().setDefault(str(R.string.login_sign_in))
+    val changeAuthButtonText = MutableLiveData<String>().setDefault(str(R.string.login_sign_up))
+    val isNewAccount = MutableLiveData<Boolean>()
+
+    val delayedAction = MutableLiveData<DelayedAction>()
+    val authProgressState = MutableLiveData<ProgressStateContainer>()
+
+    val api = Source.api
+
+    fun onAuthClicked() {
+        clickThrottler.next {
+            delayedAction.value = DelayedAction({
+                signUpOrSignIn()
+            }, 300)
+        }
     }
 
-    fun onGetStarted() {
+    fun onForgotPassword() {
         clickThrottler.next {
-            signUpOrSignIn()
+            val email = email.value
+
+            if (email.isNullOrEmpty()) {
+                authProgressState.setValue(ALERT, str(R.string.message_forgot_password_no_email))
+                return@next
+            }
+
+
+            AsyncProcessor {
+                api.sendPasswordResetEmail(email)
+                authProgressState.postValue(
+                    ALERT,
+                    str(R.string.message_forgot_password_instructions_sent)
+                )
+            }.handleError {
+                authProgressState.postValue(ALERT, "${str(R.string.message_forgot_password_error)}: $it")
+            }.run(bag)
+        }
+    }
+
+    fun toggleNewAccount() {
+        swapValues(authButtonText, changeAuthButtonText)
+
+        val newValue = isNewAccount.value != true
+        if (!newValue) {
+            delayedAction.value = DelayedAction({ confirmPassword.value = "" }, 300)
         }
+        isNewAccount.value = newValue
     }
+
+    private fun signUpOrSignIn() {
+
+        authProgressState.setValue(LOADING)
+
+        val email = email.value
+        val password = password.value
+        val confirmPassword = confirmPassword.value
+
+        val isNewAccount = isNewAccount.value
+
+        if (email.isNullOrEmpty() || password.isNullOrEmpty()) {
+            authProgressState.setValue(ALERT, str(R.string.message_auth_fields_empty))
+            return
+
+        } else if (isNewAccount == true && confirmPassword.isNullOrEmpty()) {
+            authProgressState.setValue(ALERT, str(R.string.message_auth_confirm_password))
+            return
+        } else if (isNewAccount == true && confirmPassword != password) {
+            authProgressState.setValue(ALERT, str(R.string.message_auth_passwords_do_not_match))
+            return
+        }
+
+        AsyncProcessor {
+            if (isNewAccount == true) {
+                Source.userPrefs.uuid = api.signUp(email, password)
+            } else {
+                Source.userPrefs.uuid = api.signIn(email, password)
+            }
```

Comment:
refactor smth like this:
Source.userPrefs.uuid = if (isNewAccount) api.signUp(email, password) else api.signIn(email, password)